File: Stage07-Improvements/Python/location.py
import shared
import lib.console as console

class Location:
	#constructor
	def __init__(self,
	             name:str,
	             description:str,
	             to_north:str = "",
	             to_east:str = "",
	             to_south:str = "",
	             to_west:str = "",
	             items:list[str] = [],
	             item_required:str = "",
				 enemy:str = ""):
		
		self._name:str = name
		self._description:str = description
		self._to_north:str = to_north.strip()
		self._to_east:str = to_east.strip()
		self._to_south:str = to_south.strip()
		self._to_west:str = to_west.strip()
		self._items:list[str] = items
		self._item_required:str = item_required
		self._enemy:str = enemy
		
	# properties
	@property
	def name(self) -> str:
		return self._name

	@property
	def description(self) -> str:
		return self._description

	@property
	def to_north(self) -> str:
		return self._to_north

	@property
	def to_east(self) -> str:
		return self._to_east

	@property
	def to_south(self) -> str:
		return self._to_south

	@property
	def to_west(self) -> str:
		return self._to_west
	
	@property
	def items(self) ->list:
		''' return a list of all items in this location '''
		return self._items
	
	@property
	def item_required(self) -> str:
		''' return item required to enter this location '''
		return self._item_required 
  
	# methods
	def add_item(self, item_key:str) -> None:
		''' add an item to the list of items in this location '''
		if item_key not in self._items:
			self._items.append(item_key)
	
	def remove_item(self, item_key:str) -> None:
		''' removes an item from list of items in this location '''
		if item_key in self._items:
			self._items.remove(item_key)
			
	@property
	def enemy(self):
		return self._enemy
	
	@enemy.setter
	def enemy(self, value):
		self._enemy = value	
	
	def display_location(self) -> list:
		''' descrbe the current location, any items inside it, and exits '''
		exits:list[str] = []
		if self._to_north != "":
			exits.append(f"north -> {self._to_north}") # adds name of exit to list
		if self._to_east != "":
			exits.append(f"east -> {self._to_east}")
		if self._to_south!= "":
			exits.append(f"south -> {self._to_south}")
		if self._to_west != "":
			exits.append(f"west -> {self._to_west}")
		print("".ljust(console.window_width,"─"))
		print(f"You are in a {self._name}")
		row = 2
		row += shared.print_lines(self._description, console.window_width, "","left")
		if len(exits) == 0:
			print("There are no exits")
			row += 1
		if len(self._items) > 0:
			output = "In this location there is: "
			for item in self._items:
				output += item + ", "
			output = output.strip()[:-1] # remove comma
			row += shared.print_lines(output, console.window_width, "","left")

		if self._enemy != "":
			print(f"ENEMY: {self._enemy}!")
			row += 1
		print("".ljust(console.window_width,"─"))
		row += 1
	
		return exits, row	# return exits list with name of exit as well as direction: ["east -> coridoor","north -> magic portal"]

File: Stage07-Improvements/Lua/location.lua
local Shared = require("shared")
local Console = require "lib.console"
local class = require "lib.class"
local Location = class:derive("Location")

string.PadRight = Shared.PadRight
string.Trim 	= Shared.Trim

function Location:new(name,
	             description,
	             toNorth,
	             toEast,
	             toSouth,
	             toWest,
	             items,
	             itemRequired,
	             enemy)
	self.Name 			= name
	self.Description 	= description			-- string
	self.ToNorth 		= toNorth or ""			-- string
	self.ToEast 		= toEast or ""			-- string
	self.ToSouth 		= toSouth or ""			-- string
	self.ToWest 		= toWest or ""			-- string
	self.Items 			= items or {}			-- Item keys
	self.ItemRequired 	= itemRequired or ""	-- Item key
	self.Enemy 			= enemy	or ""			-- Enemy key
end

function Location:GetName()
	return self.Name 
end

function Location:GetDescription()
	return self.Description 
end

function Location:GetToNorth()
	return self.ToNorth 
end

function Location:GetToEast()
	return self.ToEast 
end

function Location:GetToSouth()
	return self.ToSouth
end

function Location:GetToWest()
	return self.ToWest
end

function Location:GetItems()
	return self.Items
end

function Location:GetItemRequired()
	return self.ItemRequired 
end

function Location:GetEnemy()
	return self.Enemy
end

function Location:SetEnemy(value)
	self.Enemy = value
end

--[[  public methods ]]
function Location:AddItem(item) -- add item key
	table.insert(self.Items, item)
end

function Location:RemoveItem(item)
	self.Items = Shared.RemoveItem(self.Items, item)
end

function Location:DisplayLocation()
	--[[  descrbe the current location, any items inside it, and exits ]]
	local exits = {}
	if self.ToNorth ~= "" then
		table.insert(exits, "north -> "..self.ToNorth) -- adds name of exit to list
	end
	if self.ToEast ~= "" then
		table.insert(exits, "east -> "..self.ToEast) -- adds name of exit to list
	end
	if self.ToSouth ~= "" then
		table.insert(exits, "south -> "..self.ToSouth) -- adds name of exit to list
	end
	if self.ToWest ~= "" then
		table.insert(exits, "west -> "..self.ToWest) -- adds name of exit to list
	end
	print((""):PadRight(Console.WindowWidth,"─"))
	print("You are in a ".. self.Name)
	local row = 2
	row = row + Shared.PrintLines(self.Description, Console.WindowWidth, "","left")
	if #exits == 0 then
		print("There are no exits")
		row = row + 1
	end
	if #self.Items > 0 then
		local output = "In this location there is: "
		for _, item in ipairs(self.Items) do
			output = output..item .. ", "
		end
		output = output:sub(1, output:len() - 2) -- remove comma
		row = row + Shared.PrintLines(output, Console.WindowWidth, "","left")
	end
	if self.Enemy ~= "" then
		print("ENEMY: "..self.Enemy.."!")
		row = row + 1
	end

	return exits, row	-- return exits list with name of exit as well as direction: {"east -> coridoor","north -> magic portal"}
end
return Location	